File: internal/aarp/envelope.go
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

package aarp

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/luckyPipewrench/pipelock/internal/contract"
)

// Domain-separation contexts. Every signing input is the JCS canonicalization of
// an object that carries one of these as a signed field, so a signature made for
// one purpose can never be replayed as evidence for another.
const (
	// ContextAssertion is the domain separator for the assurance-assertion
	// signature (the parallel multi-signatures in Envelope.Signatures).
	ContextAssertion = "pipelock-aarp-v0.1/assurance-assertion"
)

// Envelope errors. Compare with errors.Is.
var (
	// ErrSchema means the envelope is structurally invalid: a missing required
	// field, a profile mismatch, an empty signature set, or a typed-string
	// grammar violation.
	ErrSchema = errors.New("aarp: envelope schema violation")

	// ErrUnknownField means an AARP-controlled object carried a field outside
	// the registered schema. Unlike legacy v1 receipts (where unknown fields are
	// ignored for backward compatibility), AARP objects reject unknown fields so
	// a producer cannot smuggle unsigned-but-meaningful content past appraisal.
	ErrUnknownField = errors.New("aarp: unknown field in AARP-controlled object")
)

// Subject names the immutable receipt this assurance envelope appraises, by
// digest. Every field is a typed string (number-safety): digests are lowercase
// hex, the signer key is the receipt's Ed25519 public key in hex.
type Subject struct {
	// ActionRecordSHA256 is the SHA-256 of the v1 canonical ActionRecord, or of
	// the v2 EvidenceReceipt SignablePreimage, depending on ReceiptType.
	ActionRecordSHA256 string `json:"action_record_sha256"`
	// ReceiptEnvelopeSHA256 is the SHA-256 of the canonical, unchanged receipt
	// envelope this assurance appraises.
	ReceiptEnvelopeSHA256 string `json:"receipt_envelope_sha256"`
	// ReceiptSignerKey is the hex Ed25519 public key that signed the receipt. An
	// Ed25519 public key is 32 bytes = 64 lowercase hex chars, which is why the
	// 64-hex grammar (ValidateHex256) applies; a future receipt type using a
	// different key size would need a key-length-aware validator.
	ReceiptSignerKey string `json:"receipt_signer_key"`
	// ReceiptType names which frozen receipt format the digests target. This is
	// the exact "hash-of-what" disambiguation.
	ReceiptType string `json:"receipt_type"`
}

// Receipt type constants for Subject.ReceiptType.
const (
	// ReceiptTypeActionV1 targets the legacy ActionReceipt v1 (json.Marshal +
	// SHA-256 of canonical bytes).
	ReceiptTypeActionV1 = "action_receipt_v1"
	// ReceiptTypeEvidenceV2 targets the EvidenceReceipt v2 (JCS SignablePreimage).
	ReceiptTypeEvidenceV2 = "evidence_receipt_v2"
)

var knownReceiptTypes = map[string]bool{
	ReceiptTypeActionV1:   true,
	ReceiptTypeEvidenceV2: true,
}

func (s Subject) validate() error {
	if err := ValidateHex256(s.ActionRecordSHA256); err != nil {
		return fmt.Errorf("%w: subject.action_record_sha256: %w", ErrSchema, err)
	}
	if err := ValidateHex256(s.ReceiptEnvelopeSHA256); err != nil {
		return fmt.Errorf("%w: subject.receipt_envelope_sha256: %w", ErrSchema, err)
	}
	if err := ValidateHex256(s.ReceiptSignerKey); err != nil {
		return fmt.Errorf("%w: subject.receipt_signer_key: %w", ErrSchema, err)
	}
	if !knownReceiptTypes[s.ReceiptType] {
		return fmt.Errorf("%w: unknown subject.receipt_type %q", ErrSchema, s.ReceiptType)
	}
	return nil
}

// Assertion is the producer's claim set about the subject. It is the shared
// payload bound (by digest) by every parallel signature.
type Assertion struct {
	// Claimed lists the claim names the producer asserts. They are producer
	// claims, not verified facts; the verifier confirms each one independently.
	Claimed []string `json:"claimed"`
	// MediatorID is the producer's self-declared mediator identity. It is only
	// trusted insofar as it matches a verifier-side trust entry.
	MediatorID string `json:"mediator_id"`
	// TrustDomain is the producer's SPIFFE trust domain (optional in core;
	// required when an SVID binding is present in the attestation layer).
	TrustDomain string `json:"trust_domain,omitempty"`
	// CompleteMediation is a claim-only field. v0.1 never verifies it: there is
	// no local evidence that proves the absence of an out-of-band path.
	CompleteMediation bool `json:"complete_mediation"`
	// EvidenceRefs names evidence objects (e.g. "spiffe_svid") that an upper
	// layer attaches and verifies.
	EvidenceRefs []string `json:"evidence_refs,omitempty"`
	// IssuedAt is the assertion issuance time as an RFC3339Nano typed string.
	IssuedAt string `json:"issued_at"`
}

func (a Assertion) validate() error {
	if a.MediatorID == "" {
		return fmt.Errorf("%w: assertion.mediator_id is required", ErrSchema)
	}
	if err := ValidateTimestamp(a.IssuedAt); err != nil {
		return fmt.Errorf("%w: assertion.issued_at: %w", ErrSchema, err)
	}
	if a.TrustDomain != "" {
		if err := validateTrustDomainName(a.TrustDomain); err != nil {
			return fmt.Errorf("%w: assertion.trust_domain: %w", ErrSchema, err)
		}
	}
	return nil
}

// Signature is one parallel protected signature over the shared canonical
// payload digest. Signatures are parallel (each independently binds the same
// payload digest under its own suite), never chained over one another.
type Signature struct {
	Protected ProtectedHeader `json:"protected"`
	// Sig is "<alg>:<base64-std>" over the canonical signing input.
	Sig string `json:"sig"`
}

// Envelope is the top-level AARP assurance artifact. It sits alongside a frozen
// receipt and carries its own parallel signatures; it never mutates the receipt.
type Envelope struct {
	Profile   string    `json:"profile"`
	Subject   Subject   `json:"subject"`
	Assertion Assertion `json:"assertion"`
	// Chain optionally places this envelope in an issuer's append-only,
	// hash-chained stream (Rung-1 timestamp trust). When present it is part of
	// the signed payload, so backdating within a stream is signature-detectable.
	Chain *ChainLink `json:"chain,omitempty"`
	// Signatures holds N parallel protected signatures over the payload digest.
	Signatures []Signature `json:"signatures"`
	// CritExt lists envelope-level critical extension names. Any name not in the
	// known registry rejects the whole envelope (fail closed).
	CritExt []string `json:"crit_ext"`
	// Ext carries non-critical extensions. Unknown non-critical extensions are
	// ignored safely; they are not part of the signed payload.
	Ext map[string]json.RawMessage `json:"ext,omitempty"`
}

// payload is the exact set of fields covered by every signature: the profile,
// subject, assertion, the envelope-level critical-extension list, and (when
// present) the chain link. CritExt is signed so a man-in-the-middle cannot
// strip a critical extension a producer flagged. It deliberately excludes
// Signatures (signatures never sign each other) and Ext (non-critical
// extensions are advisory and ignored safely).
type payload struct {
	Profile   string    `json:"profile"`
	Subject   Subject   `json:"subject"`
	Assertion Assertion `json:"assertion"`
	// CritExt is NOT omitempty: an empty critical-extension list must serialize
	// as "crit_ext": [] so the signed canonical bytes match an external
	// implementation that faithfully emits []. With omitempty, a nil and an
	// explicit [] would canonicalize differently across languages and break
	// cross-implementation signature agreement.
	CritExt []string   `json:"crit_ext"`
	Chain   *ChainLink `json:"chain,omitempty"`
}

func (e Envelope) payload() payload {
	// Normalize a nil critical-extension list to an empty slice so it always
	// serializes as [] (never null) in the signed payload.
	critExt := e.CritExt
	if critExt == nil {
		critExt = []string{}
	}
	return payload{
		Profile:   e.Profile,
		Subject:   e.Subject,
		Assertion: e.Assertion,
		CritExt:   critExt,
		Chain:     e.Chain,
	}
}

// CanonicalPayload returns the JCS-canonical bytes of the signed payload (the
// "identical canonical payload bytes" all parallel signatures bind). Recipe:
// marshal → ParseJSONStrict → EnforceSafeNumbers → Canonicalize.
func (e Envelope) CanonicalPayload() ([]byte, error) {
	raw, err := json.Marshal(e.payload())
	if err != nil {
		return nil, fmt.Errorf("marshal payload: %w", err)
	}
	tree, err := contract.ParseJSONStrict(raw)
	if err != nil {
		return nil, fmt.Errorf("%w: parse payload: %w", ErrSchema, err)
	}
	if err := EnforceSafeNumbers(tree); err != nil {
		return nil, err
	}
	return contract.Canonicalize(tree)
}

// PayloadDigest returns the lowercase-hex SHA-256 of the canonical payload. It
// is the value every signature binds and the value an SVID binding references
// as assurance_assertion_sha256.
func (e Envelope) PayloadDigest() (string, error) {
	canonical, err := e.CanonicalPayload()
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(canonical)
	return hex.EncodeToString(sum[:]), nil
}

// signingInput builds the canonical bytes one signature signs: the domain
// context, the shared payload digest, and that signature's protected header.
// Embedding the protected header means each signature commits to its own suite
// (alg, key, canon, profile, critical extensions), defeating algorithm
// substitution; binding the payload digest means all signatures cover identical
// payload bytes.
func signingInput(payloadDigest string, h ProtectedHeader) ([]byte, error) {
	obj := signingInputObject{
		Context:       ContextAssertion,
		PayloadSHA256: payloadDigest,
		Protected:     h,
	}
	raw, err := json.Marshal(obj)
	if err != nil {
		return nil, fmt.Errorf("marshal signing input: %w", err)
	}
	tree, err := contract.ParseJSONStrict(raw)
	if err != nil {
		return nil, fmt.Errorf("parse signing input: %w", err)
	}
	return contract.Canonicalize(tree)
}

type signingInputObject struct {
	Context       string          `json:"context"`
	PayloadSHA256 string          `json:"payload_sha256"`
	Protected     ProtectedHeader `json:"protected"`
}

// validatePayloadParts validates the signed payload-bearing fields (profile,
// subject, assertion, chain) plus envelope-level critical extensions. It does
// NOT require signatures, so it is the structural gate the signer runs before
// producing them.
func (e Envelope) validatePayloadParts() error {
	if e.Profile != Profile {
		return fmt.Errorf("%w: profile %q, want %q", ErrSchema, e.Profile, Profile)
	}
	if err := e.Subject.validate(); err != nil {
		return err
	}
	if err := e.Assertion.validate(); err != nil {
		return err
	}
	if e.Chain != nil {
		if err := e.Chain.validate(); err != nil {
			return err
		}
	}
	return checkCriticalExtensions(e.CritExt)
}

// validateStructure runs the envelope-fatal schema checks that must hold before
// any per-signature appraisal: the payload parts (which include the top-level
// profile and the envelope-level critical-extension list) and a non-empty
// signature set.
//
// Only fields inside the SIGNED payload are envelope-fatal here. Per-signature
// suite fields (a signature's protected profile, canon, alg, key trust, and its
// own critical-extension list) are deliberately NOT checked here: they are
// per-signature outcomes appraised in appraiseSignature. The signatures array
// is not itself signed, so a man-in-the-middle can append a junk signature; if a
// bad protected header were envelope-fatal, that append would deny a
// legitimately-signed envelope. Per-signature handling makes one unverifiable
// signature inert instead of fatal, while the signed top-level profile and
// crit_ext (which an append cannot forge) stay fatal.
func (e Envelope) validateStructure() error {
	if err := e.validatePayloadParts(); err != nil {
		return err
	}
	if len(e.Signatures) == 0 {
		return fmt.Errorf("%w: envelope has no signatures", ErrSchema)
	}
	return nil
}

// Marshal returns the JSON encoding of an envelope. An empty critical-extension
// list is emitted as "crit_ext": [] (never omitted), matching the signed payload
// and the spec wire form so strict external implementations and golden vectors
// agree byte-for-byte.
func Marshal(e Envelope) ([]byte, error) {
	if e.CritExt == nil {
		e.CritExt = []string{}
	}
	return json.Marshal(e)
}

// Unmarshal parses a JSON-encoded envelope, rejecting duplicate keys at any
// nesting depth (parser-differential smuggling guard) and unknown fields in
// AARP-controlled objects. The receipt subject digests, not the receipt bytes,
// are what AARP signs, so strict decoding here cannot affect frozen receipts.
func Unmarshal(data []byte) (Envelope, error) {
	tree, err := contract.ParseJSONStrict(data)
	if err != nil {
		return Envelope{}, fmt.Errorf("%w: %w", ErrSchema, err)
	}
	// Reject any raw JSON number outside the I-JSON safe range anywhere in the
	// envelope before decoding. Identity, digest, counter, timestamp, and amount
	// fields must arrive as typed strings; a raw number in any position is a
	// cross-language canonicalization hazard and is refused here.
	if err := EnforceSafeNumbers(tree); err != nil {
		return Envelope{}, err
	}
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.DisallowUnknownFields()
	dec.UseNumber() // future-proof: preserve integers if a numeric field is ever added
	var e Envelope
	if err := dec.Decode(&e); err != nil {
		// Distinguish an unknown AARP field (a smuggling attempt) from an
		// ordinary type/shape decode error, so errors.Is(err, ErrUnknownField)
		// is not a false positive on, e.g., a type mismatch.
		if strings.Contains(err.Error(), "unknown field") {
			return Envelope{}, fmt.Errorf("%w: %w", ErrUnknownField, err)
		}
		return Envelope{}, fmt.Errorf("%w: %w", ErrSchema, err)
	}
	return e, nil
}

File: sdk/verifiers/ts/src/aarp/envelope.ts
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

// Envelope schema and signed-payload construction, ported from
// internal/aarp/envelope.go (+ chain.go). The strict tree from parseJSONStrict
// is decoded into typed structures here, rejecting unknown fields in
// AARP-controlled objects so a producer cannot smuggle unsigned content past
// appraisal. ext is a free map and is NOT part of the signed payload.

import { createHash } from "node:crypto";
import { canonicalize } from "./canonical.js";
import {
  GrammarError,
  validateHex256,
  validateTimestamp,
  validateUint64String,
} from "./numbers.js";
import { RawNumber } from "./strictjson.js";
import { PROFILE, ProtectedHeader, checkCriticalExtensions } from "./suite.js";

// ContextAssertion is the domain separator for the assurance-assertion signature.
export const CONTEXT_ASSERTION = "pipelock-aarp-v0.1/assurance-assertion";

// GenesisPriorHash is the prior-hash sentinel for the first link in a stream.
export const GENESIS_PRIOR_HASH = "0".repeat(64);

// SchemaError marks an envelope-fatal schema violation (missing/extra field,
// wrong type, bad grammar). The CLI maps it to exit 1.
export class SchemaError extends Error {
  readonly code = 1;
}

const knownReceiptTypes: Record<string, boolean> = {
  action_receipt_v1: true,
  evidence_receipt_v2: true,
};

export interface Subject {
  action_record_sha256: string;
  receipt_envelope_sha256: string;
  receipt_signer_key: string;
  receipt_type: string;
}

export interface Assertion {
  claimed: string[];
  mediator_id: string;
  trust_domain?: string;
  complete_mediation: boolean;
  evidence_refs?: string[];
  issued_at: string;
}

export interface ChainLink {
  issuer_id: string;
  seq: string;
  prior_hash: string;
}

export interface Signature {
  protected: ProtectedHeader;
  sig: string;
}

export interface Envelope {
  profile: string;
  subject: Subject;
  assertion: Assertion;
  chain?: ChainLink;
  signatures: Signature[];
  crit_ext?: string[];
}

// asObject narrows a strict-tree value to a plain object or throws schema-fatal.
function asObject(v: unknown, field: string): Record<string, unknown> {
  if (typeof v !== "object" || v === null || Array.isArray(v) || v instanceof RawNumber) {
    throw new SchemaError(`${field} must be an object`);
  }
  return v as Record<string, unknown>;
}

function asString(v: unknown, field: string): string {
  if (typeof v !== "string") throw new SchemaError(`${field} must be a string`);
  return v;
}

function asBool(v: unknown, field: string): boolean {
  if (typeof v !== "boolean") throw new SchemaError(`${field} must be a boolean`);
  return v;
}

function asStringArray(v: unknown, field: string): string[] {
  if (!Array.isArray(v)) throw new SchemaError(`${field} must be an array`);
  return v.map((item, i) => asString(item, `${field}[${i}]`));
}

// rejectUnknownKeys enforces that an AARP-controlled object carries only the
// allowed fields, mirroring Go's DisallowUnknownFields on the typed structs.
function rejectUnknownKeys(
  obj: Record<string, unknown>,
  allowed: readonly string[],
  field: string,
): void {
  const allow = new Set(allowed);
  for (const key of Object.keys(obj)) {
    if (!allow.has(key)) {
      throw new SchemaError(`unknown field ${JSON.stringify(key)} in ${field}`);
    }
  }
}

function decodeSubject(v: unknown): Subject {
  const obj = asObject(v, "subject");
  rejectUnknownKeys(
    obj,
    ["action_record_sha256", "receipt_envelope_sha256", "receipt_signer_key", "receipt_type"],
    "subject",
  );
  return {
    action_record_sha256: asString(obj.action_record_sha256, "subject.action_record_sha256"),
    receipt_envelope_sha256: asString(
      obj.receipt_envelope_sha256,
      "subject.receipt_envelope_sha256",
    ),
    receipt_signer_key: asString(obj.receipt_signer_key, "subject.receipt_signer_key"),
    receipt_type: asString(obj.receipt_type, "subject.receipt_type"),
  };
}

function decodeAssertion(v: unknown): Assertion {
  const obj = asObject(v, "assertion");
  rejectUnknownKeys(
    obj,
    ["claimed", "mediator_id", "trust_domain", "complete_mediation", "evidence_refs", "issued_at"],
    "assertion",
  );
  const out: Assertion = {
    claimed: obj.claimed === undefined ? [] : asStringArray(obj.claimed, "assertion.claimed"),
    mediator_id: asString(obj.mediator_id, "assertion.mediator_id"),
    complete_mediation:
      obj.complete_mediation === undefined
        ? false
        : asBool(obj.complete_mediation, "assertion.complete_mediation"),
    issued_at: asString(obj.issued_at, "assertion.issued_at"),
  };
  if (obj.trust_domain !== undefined) {
    out.trust_domain = asString(obj.trust_domain, "assertion.trust_domain");
  }
  if (obj.evidence_refs !== undefined) {
    out.evidence_refs = asStringArray(obj.evidence_refs, "assertion.evidence_refs");
  }
  return out;
}

function decodeChainLink(v: unknown): ChainLink {
  const obj = asObject(v, "chain");
  rejectUnknownKeys(obj, ["issuer_id", "seq", "prior_hash"], "chain");
  return {
    issuer_id: asString(obj.issuer_id, "chain.issuer_id"),
    seq: asString(obj.seq, "chain.seq"),
    prior_hash: asString(obj.prior_hash, "chain.prior_hash"),
  };
}

function decodeProtected(v: unknown, idx: number): ProtectedHeader {
  const obj = asObject(v, `signatures[${idx}].protected`);
  rejectUnknownKeys(
    obj,
    ["profile", "canon", "alg", "key_type", "key_id", "signer_role", "crit"],
    `signatures[${idx}].protected`,
  );
  const header: ProtectedHeader = {
    profile: asString(obj.profile, `signatures[${idx}].protected.profile`),
    canon: asString(obj.canon, `signatures[${idx}].protected.canon`),
    alg: asString(obj.alg, `signatures[${idx}].protected.alg`),
    key_type: asString(obj.key_type, `signatures[${idx}].protected.key_type`),
    key_id: asString(obj.key_id, `signatures[${idx}].protected.key_id`),
    signer_role: asString(obj.signer_role, `signatures[${idx}].protected.signer_role`),
  };
  if (obj.crit !== undefined) {
    header.crit = asStringArray(obj.crit, `signatures[${idx}].protected.crit`);
  }
  return header;
}

function decodeSignature(v: unknown, idx: number): Signature {
  const obj = asObject(v, `signatures[${idx}]`);
  rejectUnknownKeys(obj, ["protected", "sig"], `signatures[${idx}]`);
  if (obj.protected === undefined) {
    throw new SchemaError(`signatures[${idx}].protected is required`);
  }
  return {
    protected: decodeProtected(obj.protected, idx),
    sig: asString(obj.sig, `signatures[${idx}].sig`),
  };
}

// decodeEnvelope decodes a strict-parsed tree into a typed Envelope, rejecting
// unknown fields in every AARP-controlled object. ext is allowed and ignored.
export function decodeEnvelope(tree: unknown): Envelope {
  const obj = asObject(tree, "envelope");
  rejectUnknownKeys(
    obj,
    ["profile", "subject", "assertion", "chain", "signatures", "crit_ext", "ext"],
    "envelope",
  );
  if (obj.subject === undefined) throw new SchemaError("subject is required");
  if (obj.assertion === undefined) throw new SchemaError("assertion is required");
  if (obj.signatures === undefined || !Array.isArray(obj.signatures)) {
    throw new SchemaError("signatures must be an array");
  }
  const env: Envelope = {
    profile: asString(obj.profile, "profile"),
    subject: decodeSubject(obj.subject),
    assertion: decodeAssertion(obj.assertion),
    signatures: obj.signatures.map((s, i) => decodeSignature(s, i)),
  };
  if (obj.chain !== undefined) {
    env.chain = decodeChainLink(obj.chain);
  }
  if (obj.crit_ext !== undefined) {
    env.crit_ext = asStringArray(obj.crit_ext, "crit_ext");
  }
  return env;
}

function wrapSchema(err: unknown): never {
  if (err instanceof GrammarError) throw new SchemaError(err.message);
  throw err;
}

function validateSubject(s: Subject): void {
  try {
    validateHex256(s.action_record_sha256);
  } catch (err) {
    wrapSchema(err);
  }
  try {
    validateHex256(s.receipt_envelope_sha256);
  } catch (err) {
    wrapSchema(err);
  }
  try {
    validateHex256(s.receipt_signer_key);
  } catch (err) {
    wrapSchema(err);
  }
  if (!knownReceiptTypes[s.receipt_type]) {
    throw new SchemaError(`unknown subject.receipt_type ${JSON.stringify(s.receipt_type)}`);
  }
}

function validateAssertion(a: Assertion): void {
  if (a.mediator_id === "") {
    throw new SchemaError("assertion.mediator_id is required");
  }
  try {
    validateTimestamp(a.issued_at);
  } catch (err) {
    wrapSchema(err);
  }
  // trust_domain validation is intentionally minimal in core; an empty string is
  // simply absent. The Go core validates a SPIFFE name shape only when set, but
  // the corpus never exercises that path, so a non-empty value is accepted here
  // as the Go core would accept any well-formed SPIFFE-like string.
}

function validateChainLink(c: ChainLink): void {
  if (c.issuer_id === "") {
    throw new SchemaError("chain.issuer_id is required");
  }
  try {
    validateUint64String(c.seq);
  } catch (err) {
    wrapSchema(err);
  }
  try {
    validateHex256(c.prior_hash);
  } catch (err) {
    wrapSchema(err);
  }
  if (c.seq === "0" && c.prior_hash !== GENESIS_PRIOR_HASH) {
    throw new SchemaError("genesis link (seq 0) must carry the zero prior hash");
  }
  if (c.seq !== "0" && c.prior_hash === GENESIS_PRIOR_HASH) {
    throw new SchemaError(`non-genesis link (seq ${c.seq}) must not carry the genesis prior hash`);
  }
}

// validatePayloadParts validates the signed payload-bearing fields plus
// envelope-level critical extensions.
export function validatePayloadParts(e: Envelope): void {
  if (e.profile !== PROFILE) {
    throw new SchemaError(`profile ${JSON.stringify(e.profile)}, want ${JSON.stringify(PROFILE)}`);
  }
  validateSubject(e.subject);
  validateAssertion(e.assertion);
  if (e.chain !== undefined) {
    validateChainLink(e.chain);
  }
  checkCriticalExtensions(e.crit_ext);
}

// validateStructure runs the envelope-fatal schema checks before any
// per-signature appraisal: the payload parts (which include the top-level
// profile and the envelope-level critical-extension list) and a non-empty
// signature set.
//
// Only fields inside the SIGNED payload are envelope-fatal here. Per-signature
// suite fields (a signature's protected profile, canon, and its own
// critical-extension list) are deliberately NOT checked here: they are
// per-signature outcomes appraised in appraiseSignature. The signatures array
// is not itself signed, so a man-in-the-middle can append a junk signature; if
// a bad protected header were envelope-fatal, that append would deny a
// legitimately-signed envelope. Per-signature handling makes one unverifiable
// signature inert instead of fatal, while the signed top-level profile and
// crit_ext (which an append cannot forge) stay fatal.
export function validateStructure(e: Envelope): void {
  validatePayloadParts(e);
  if (e.signatures.length === 0) {
    throw new SchemaError("envelope has no signatures");
  }
}

// payloadObject builds the JCS-input object for the signed payload: profile,
// subject, assertion, crit_ext, optional chain. It excludes signatures and ext,
// and applies Go's omitempty semantics for the optional assertion fields so the
// canonical bytes match the Go marshaller exactly. crit_ext is NOT omitempty: a
// nil/absent envelope crit_ext must serialize as "crit_ext": [] (never omitted,
// never null), matching Go's payload() which normalizes nil to []string{} with a
// no-omitempty struct tag. Without this, the computed payload digest differs
// from Go and every signature fails to verify.
function payloadObject(e: Envelope): Record<string, unknown> {
  const assertion: Record<string, unknown> = {
    claimed: e.assertion.claimed,
    mediator_id: e.assertion.mediator_id,
    complete_mediation: e.assertion.complete_mediation,
    issued_at: e.assertion.issued_at,
  };
  if (e.assertion.trust_domain !== undefined && e.assertion.trust_domain !== "") {
    assertion.trust_domain = e.assertion.trust_domain;
  }
  if (e.assertion.evidence_refs !== undefined && e.assertion.evidence_refs.length > 0) {
    assertion.evidence_refs = e.assertion.evidence_refs;
  }
  const payload: Record<string, unknown> = {
    profile: e.profile,
    subject: {
      action_record_sha256: e.subject.action_record_sha256,
      receipt_envelope_sha256: e.subject.receipt_envelope_sha256,
      receipt_signer_key: e.subject.receipt_signer_key,
      receipt_type: e.subject.receipt_type,
    },
    assertion,
  };
  payload.crit_ext = e.crit_ext === undefined ? [] : e.crit_ext;
  if (e.chain !== undefined) {
    payload.chain = {
      issuer_id: e.chain.issuer_id,
      seq: e.chain.seq,
      prior_hash: e.chain.prior_hash,
    };
  }
  return payload;
}

// canonicalPayload returns the JCS-canonical bytes of the signed payload.
export function canonicalPayload(e: Envelope): Buffer {
  return Buffer.from(canonicalize(payloadObject(e)), "utf8");
}

// payloadDigest returns the lowercase-hex SHA-256 of the canonical payload.
export function payloadDigest(e: Envelope): string {
  return createHash("sha256").update(canonicalPayload(e)).digest("hex");
}

// signingInput builds the canonical bytes one signature signs: the domain
// context, the shared payload digest, and that signature's protected header.
export function signingInput(payloadSha256: string, h: ProtectedHeader): Buffer {
  const protectedObj: Record<string, unknown> = {
    profile: h.profile,
    canon: h.canon,
    alg: h.alg,
    key_type: h.key_type,
    key_id: h.key_id,
    signer_role: h.signer_role,
  };
  // Go marshals crit with omitempty: present only when the slice is non-empty.
  if (h.crit !== undefined && h.crit.length > 0) {
    protectedObj.crit = h.crit;
  }
  const obj = {
    context: CONTEXT_ASSERTION,
    payload_sha256: payloadSha256,
    protected: protectedObj,
  };
  return Buffer.from(canonicalize(obj), "utf8");
}
